### DISCUSSION TITLE
Is there a package manager for Grain?

### DISCUSSION THREAD

**Discussion by matthew-dean** [grain-lang/grain#1789] [General]
I was thinking about making a parser, but first making something like a simple parser combinator library, but where are Grain packages published, and how are they referenced in code?

**Comment by spotandjake** [comment]
There isn't really a place where they are published yet the idea will eventually be a package manager named `silo` but that's a bit out atm. some packages are being added to the awesome-grain repo under the grain org though.

**Comment by makoConstruct** [comment]
Are you familiar with Nix? It's a simple functional programming language, and is also, afaik, the only cross-language (cross-platform) package manager and build language. It's a really important thing, but it has problems. It was built before wasm, so has a lot of awkward platform-specific builds. It's not a real programming language (_has no type analysis, so terrible error messages_), and it has an unintuitive syntax, and there was a clear mistake made in the array syntax. It also doesn't integrate with content-addressed distributed storage, which is bad for reproducibility and makes everyone reliant on the official nix cache, or, worse, on github.

So I've actually been considering Grain + IPFS/IPVM as the next incarnation of Nix, and if you could point me to a wasm build of the Grain compiler, I may start building a cross-language build system/package manager out of that very soon.
Is there a wasm build?

**Reply by coder0107git** [reply]
There currently isn't a wasm build (afaik) of the grain compiler. The compiler is actually OCaml compiled to javascript, and bundled with the command line written in nodejs as a self contained native executable. All of the OCaml to WASM compilers that I have looked at are experimental (however [this one](https://github.com/ocaml-wasm/wasm_of_ocaml) looks promising). You could try compiling the compiler to plain browser javascript and run that inside of a javascript engine compiled to a wasm (e.g. QuickJS).
